storage/fanout.go

```diff
@@ -66,23 +74,26 @@ func (f *fanout) Querier(ctx context.Context, mint, maxt int64) (Querier, error)
 	queriers := make([]Querier, 0, 1+len(f.secondaries))
 
 	// Add primary querier
-	querier, err := f.primary.Querier(ctx, mint, maxt)
+	primaryQuerier, err := f.primary.Querier(ctx, mint, maxt)
 	if err != nil {
 		return nil, err
 	}
-	queriers = append(queriers, querier)
+	queriers = append(queriers, primaryQuerier)
 
 	// Add secondary queriers
 	for _, storage := range f.secondaries {
 		querier, err := storage.Querier(ctx, mint, maxt)
+		if mergeQuerier, ok := querier.(*mergeQuerier); ok {
```

Comment:
I needed a reference from the querier back to the fanout storage so that if the querier had a remote error, it could notify the fanout storage and the fanout would remember which queriers had remote errors.

Reply:
What if the underlying querier isn't a mergeQuerier though?

Reply:
Then it shouldn't have to worry about knowing the storage source. What I found was that if theres >1 remote storage source, then those remote storages get combined into a mergeQuerier, which is nested inside of another mergeQuerier that contains the querier for the remote sources and the querier for the primary source. So I wanted a reference from that mergeQuerier back to the fanout storage too, so that those remote errors would get remembered.

The alternative is that there's 1 remote storage, and that isn't merged into a mergeQuerier. In that case, remote storage errors are identified in the top level mergeQuerier already so there's already a reference to the fanout storage. I hope that makes sense